orttraining/orttraining/eager/ort_aten.cpp

```diff
@@ -937,6 +941,103 @@ at::Tensor& abs_out(
   return out;
 }
 
+// eq has to cast the output to align with expected aten behavior
+
+// aten::eq.Tensor_out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
+Tensor& eq_Tensor_out(
+  const Tensor& self,
+  const Tensor& other,
+  // *,
+  Tensor& out) {
+  ORT_LOG_FN(self, other, out);
+
+  if (
+    !IsSupportedType(self, {at::kBool,at::kLong,at::kHalf,at::kByte,at::kShort,at::kDouble,at::kBFloat16,at::kFloat,at::kInt}) ||
+    !IsSupportedType(other, {at::kBool,at::kLong,at::kHalf,at::kByte,at::kShort,at::kDouble,at::kBFloat16,at::kFloat,at::kInt})) {
+    return native::call_fallback_fn<
+      &native::cpu_fallback,
+      ATEN_OP(eq_Tensor_out)>::call(self, other, out);
+  }
+  auto& invoker = GetORTInvoker(self.device());
+
+  // resize the output and then create output ort value to be updated.
+  resize_output(invoker, dynamic_cast<ORTTensorImpl*>(out.unsafeGetTensorImpl()), self.sizes());
+  auto ort_input_out = create_ort_value(invoker, out);
+
+  auto ort_input_self = create_ort_value(invoker, self);
+  auto ort_input_other = create_ort_value(invoker, other);
+
+  std::vector<OrtValue> ort_outputs_0_Equal(1);
+
+  auto status = invoker.Invoke("Equal", {
+    std::move(ort_input_self),
+    std::move(ort_input_other),
+  }, ort_outputs_0_Equal, nullptr);
+
+  if (!status.IsOK())
+    throw std::runtime_error(
+      "ORT return failure status:" + status.ErrorMessage());
+
+  auto temp = CastToType(invoker, ort_outputs_0_Equal[0], out.scalar_type());
```

Comment:
One downside is that in (what I expect to be) the common case, which out is a tensor of type `bool`, we will have an extra allocation, copy, and cast sequence...

Reply:
if we think that is a big downside, we can use the equal_onnxop_found flag to add some code to avoid the copy. I have an idea of how to do this.

Reply:
see last commit I pushed which addresses this (https://github.com/microsoft/onnxruntime/pull/12031/commits/c169b380af3501d4052ca20de0f7ade3ce3c566c)